js.go

```diff
@@ -0,0 +1,1175 @@
+}
+
+// ApiError is included in all API responses if there was an error.
+type ApiError struct {
+	Code        int    `json:"code"`
+	Description string `json:"description,omitempty"`
+}
+
+// ApiResponse is a standard response from the JetStream JSON API
+type ApiResponse struct {
+	Type  string    `json:"type"`
+	Error *ApiError `json:"error,omitempty"`
+}
+
+type AccountInfoResponse struct {
+	ApiResponse
+	*AccountStats
+}
+
+// AccountLimits is for the information about
+type AccountLimits struct {
+	MaxMemory    int64 `json:"max_memory"`
+	MaxStore     int64 `json:"max_storage"`
+	MaxStreams   int   `json:"max_streams"`
+	MaxConsumers int   `json:"max_consumers"`
+}
+
+// AccountStats returns current statistics about the account's JetStream usage.
+type AccountStats struct {
+	Memory  uint64        `json:"memory"`
+	Store   uint64        `json:"storage"`
+	Streams int           `json:"streams"`
+	Limits  AccountLimits `json:"limits"`
+}
+
+const defaultRequestWait = 5 * time.Second
+
+// Internal struct for jetstream
+type js struct {
+	nc *Conn
+	// For import JS from other accounts.
+	pre string
+}
+
+// Request API subjects for JetStream.
+const (
+	// JSApiAccountInfo is for obtaining general information about JetStream.
+	JSApiAccountInfo = "$JS.API.INFO"
+	// JSApiStreams can lookup a stream by subject.
+	JSApiStreams = "$JS.API.STREAM.NAMES"
+	// JSApiConsumerCreateT is used to create consumers.
+	JSApiConsumerCreateT = "$JS.API.CONSUMER.CREATE.%s"
+	// JSApiDurableCreateT is used to create durable consumers.
+	JSApiDurableCreateT = "$JS.API.CONSUMER.DURABLE.CREATE.%s.%s"
+	// JSApiConsumerInfoT is used to create consumers.
+	JSApiConsumerInfoT = "$JS.API.CONSUMER.INFO.%s.%s"
+	// JSApiRequestNextT is the prefix for the request next message(s) for a consumer in worker/pull mode.
+	JSApiRequestNextT = "$JS.API.CONSUMER.MSG.NEXT.%s.%s"
+	// JSApiStreamCreate is the endpoint to create new streams.
+	JSApiStreamCreateT = "$JS.API.STREAM.CREATE.%s"
+)
+
+// JetStream returns a JetStream context for pub/sub interactions.
+func (nc *Conn) JetStream(optApiPre ...string) (JetStream, error) {
+	js := &js{nc: nc}
+
+	if len(optApiPre) > 0 {
+		const tokSep = "."
+		if len(optApiPre) == 1 && len(optApiPre[0]) > 0 {
+			js.pre = optApiPre[0]
+			if !strings.HasSuffix(js.pre, tokSep) {
+				js.pre = js.pre + tokSep
+			}
+		} else {
+			return nil, ErrJetStreamBadPre
+		}
+	}
+
+	resp, err := nc.Request(js.apiSubj(JSApiAccountInfo), nil, defaultRequestWait)
+	if err != nil {
+		if err == ErrNoResponders {
+			err = ErrJetStreamNotEnabled
+		}
+		return nil, err
+	}
+	var info AccountInfoResponse
+	if err := json.Unmarshal(resp.Data, &info); err != nil {
+		return nil, err
+	}
+	if info.Error != nil && info.Error.Code == 503 {
+		return nil, ErrJetStreamNotEnabled
+	}
+	return js, nil
+}
+
+func (js *js) apiSubj(subj string) string {
+	if js.pre == _EMPTY_ {
+		return subj
+	}
+	var b strings.Builder
+	b.WriteString(js.pre)
+	b.WriteString(subj)
```

Comment:
Is the intention that people would import like `js.derek.$JS.API.....` since thats all this supports.

I'd have thought thats not how one would do it, ie. you export `$JS.API.>` and I import that into `js.derek` leaving subjects like `js.derek.INFO`

Reply:
That would involve full on mapping vs just prefix prepend. But I think we want to match what folks would want to do here from the application space.

Reply:
Actually I misspoke, that is possible. Will make that change and the func options for the JetStream context itself.

Reply:
ok this should work and the import test works now properly.